vm/isolated/isolated.go

```diff
@@ -189,39 +199,57 @@ func (inst *instance) ssh(command string) error {
 	return nil
 }
 
+func (inst *instance) waitRebootAndSSH(rebootTimeout int, sshTimeout time.Duration) error {
+	if err := inst.waitForReboot(rebootTimeout); err != nil {
+		log.Logf(2, "isolated: machine did not reboot")
+		return err
+	}
+	log.Logf(2, "isolated: rebooted wait for comeback")
+	if err := inst.waitForSSH(sshTimeout); err != nil {
+		log.Logf(2, "isolated: machine did not comeback")
+		return err
+	}
+	log.Logf(2, "isolated: reboot succeeded")
+	return nil
+}
+
 func (inst *instance) repair() error {
 	log.Logf(2, "isolated: trying to ssh")
-	if err := inst.waitForSSH(30 * time.Minute); err == nil {
-		if inst.cfg.TargetReboot {
-			if len(inst.cfg.USBDevNums) > 0 {
-				log.Logf(2, "isolated: trying to reboot by USB authorization")
-				usbAuth := fmt.Sprintf("%s%s%s", "/sys/bus/usb/devices/", inst.cfg.USBDevNums[inst.index], "/authorized")
-				if err := ioutil.WriteFile(usbAuth, []byte("0"), 0); err != nil {
-					log.Logf(2, "isolated: failed to turn off the device")
-					return err
-				}
-				if err := ioutil.WriteFile(usbAuth, []byte("1"), 0); err != nil {
-					log.Logf(2, "isolated: failed to turn on the device")
-					return err
-				}
-			} else {
-				log.Logf(2, "isolated: ssh succeeded, trying to reboot by ssh")
-				inst.ssh("reboot") // reboot will return an error, ignore it
+	if err := inst.waitForSSH(30 * time.Minute); err != nil {
+		log.Logf(2, "isolated: ssh failed")
+		return fmt.Errorf("SSH failed")
+	}
+	if inst.cfg.TargetReboot {
+		if len(inst.cfg.USBDevNums) > 0 {
+			log.Logf(2, "isolated: trying to reboot by USB authorization")
+			usbAuth := fmt.Sprintf("%s%s%s", "/sys/bus/usb/devices/", inst.cfg.USBDevNums[inst.index], "/authorized")
+			if err := ioutil.WriteFile(usbAuth, []byte("0"), 0); err != nil {
+				log.Logf(2, "isolated: failed to turn off the device")
+				return err
+			}
+			if err := ioutil.WriteFile(usbAuth, []byte("1"), 0); err != nil {
+				log.Logf(2, "isolated: failed to turn on the device")
+				return err
 			}
+		} else {
+			log.Logf(2, "isolated: ssh succeeded, trying to reboot by ssh")
+			inst.ssh("reboot") // reboot will return an error, ignore it
 		}
-		if err := inst.waitForReboot(5 * 60); err != nil {
-			log.Logf(2, "isolated: machine did not reboot")
-			return err
+	}
+	if err := inst.waitRebootAndSSH(5*60, 30*time.Minute); err != nil {
+		return fmt.Errorf("waitRebootAndSSH failed: %v", err)
+	}
+	if inst.cfg.StartupScript != "" {
+		log.Logf(2, "isolated: executing startup_script")
+		// Execute the contents of the StartupScript on the DUT.
+		contents, err := ioutil.ReadFile(inst.cfg.StartupScript)
+		if err != nil {
+			return fmt.Errorf("unable to read startup_script: %v", err)
 		}
-		log.Logf(2, "isolated: rebooted wait for comeback")
-		if err := inst.waitForSSH(30 * time.Minute); err != nil {
-			log.Logf(0, "isolated: machine did not comeback")
-			return err
+		if err := inst.ssh(fmt.Sprintf("bash -c '%v'", string(contents))); err != nil {
```

Comment:
Can this script contain new lines?
But if it contains a single quote, I guess it will end the bash command. Is it possible to escape quotes with \?

Reply:
New lines don't seem to be a problem, but escaping nested single quotes does not seem to be supported in bash(https://www.gnu.org/software/bash/manual/html_node/Double-Quotes.html). We could use `"%v"` and escape all double quotes(and double-quotes-escapes) already inside the script file. With that a script as follows works :

```
touch \
    /tmp/OK
touch '/tmp/OK2'
touch "/tmp/OK3"
touch /tmp/OK4
bash -c "bash -c \"ls -al\"" > OK5
```

Reply:
Also, this means that we need to not escape other special characters(and escapes of those which might exist). We are also assuming bash is on the DUT. Is this method a bit fragile?
I wonder if it would be better to have a configuration that allows for remounting /tmp(or any specified mountpoint) on the DUT as "exec", and then support copying the script over to that directory on the DUT and executing it. Does that make sense?

Reply:
I think ssh already starts a shell. So some shell should be there.
If that will not work on some machines, perhaps we could query shell from SHELL env var.

Reply:
Using `"%v"` looks good, but I think we need to escape " in vm/isolated.
Asking for a script that is escaped looks like a wrong design for the feature. What is this format? Its't not used/asked by anything else. If it's not escaped a user may run it manually for testing. If it's escaped it can't be run manually anymore.

Reply:
I think I wasn't clearer earlier; let me try again. My concern was that the script might already contain escaped double-quotes(I'm not suggesting that user should provide us a quoted script). For eg, the user might have already used nested double quotes in his script for doing something like :

```
bash -c "bash -c \"ls -al\"" > OK5
```
Now when `vm/isolated` tries to execute this, if we are attempting to escape the `"`, we will also need to escape the `\"` that already exists in the script file.

Reply:
You are right. Is there a way to escape `\"` as well?